src/repair.rs

```diff
@@ -93,164 +100,223 @@ impl<N: Network> Repair<N> {
     ) -> Self {
         let validators = epoch_info.validators.clone();
         let sampler = StakeWeightedSampler::new(validators);
-        let repair_channel = (repair_channel.0, tokio::sync::Mutex::new(repair_channel.1));
+        let repair_channel = (repair_channel.0, Some(repair_channel.1));
         Self {
             blockstore,
             pool,
+
+            slice_roots: BTreeMap::new(),
+            outstanding_requests: BTreeMap::new(),
+            request_timeouts: BinaryHeap::new(),
+
             network,
             repair_channel,
             sampler,
+            epoch_info,
         }
     }
 
-    /// The main repair loop.
+    /// Main loop of the repair protocol.
     ///
     /// Listens to incoming requests for blocks to repair on `self.repair_channel`.
     /// Inititates the corresponding repair process and handles ongoing repairs.
-    pub async fn repair_loop(&self) {
-        // only this loop uses the receiver
-        let mut guard = self.repair_channel.1.lock().await;
-        while let Some((slot, hash)) = guard.recv().await {
-            self.repair_block(slot, hash).await;
+    /// Also, listens to incoming repair messages from the network and handles them.
+    pub async fn repair_loop(&mut self) {
+        let mut repair_receiver = self.repair_channel.1.take().unwrap();
+        loop {
+            let next_timeout = self.request_timeouts.peek().map(|t| t.0);
+            let sleep_duration = match next_timeout {
+                None => std::time::Duration::MAX,
+                Some(t) => t.duration_since(Instant::now()),
+            };
+            tokio::select! {
+                // handle repair request or response from network
+                res = self.receive() => self.handle_repair_message(res.unwrap()).await.unwrap(),
+                // handle request for repairing new block
```

Comment:
```suggestion
                // handle request from other components in the node for repairing new block
```

Because we are inside a macro, my IDE is not giving me type hints and it is not super clear which arm is handling network messages and which one is handling messages on a channel.  

Not sure if my suggestion above helps improve things.  Maybe you have another idea.

Reply:
Another suggestion is to put in some explicit type annotations to help the reader.

Reply:
that seems to be an IDE/setup specific problem, I can see types w/o problem here